### DISCUSSION TITLE
how to typing column enu in the model?

### DISCUSSION THREAD

**Discussion by kkkasio** [adonisjs/core#1805] [Help]
Hello dear friends!
I have a question when defining the columns within my model

In my database I want to have an enum type field that contains some values.
When I am going to map this to my model, I know that it should logically pass as a "listing" so she can know what value to accept (as if she were typing a list)
my question is if i just pass string and keep happy, or if there really is a possibility to type the values.

there is nothing said in the Columns documentation on the subject.
https://preview.adonisjs.com/guides/models/columns#defining-columns
Thanks so much :)

![Captura de tela de 2020-10-28 00-02-09](https://user-images.githubusercontent.com/19254402/97385766-3573ed80-18b1-11eb-8f59-a10dc6325e40.png)

**Comment by arthurcvm** [comment]
Hi @kkkasio, you discover the solution? I have same problem...

**Comment by iagobruno** [comment]
Here's my solution:

First I create an [enum ](https://www.typescriptlang.org/docs/handbook/enums.html)with all possible values:

```ts
// ./contracts/enums.ts
export enum PostStatus {
  DRAFT = 'draft',
  PUBLIC = 'public',
  PRIVATE = 'private',
  DELETED = 'deleted',
}
```

Next, I import the enum into migration file and use `Object.values()` to transform it into an array:

```ts
// ./database/migrations/1601605294173_create_posts_table.ts
import BaseSchema from '@ioc:Adonis/Lucid/Schema'
import { PostStatus } from 'Contracts/enums'

export default class extends BaseSchema {
  protected tableName = 'posts'

  public async up () {
    this.schema.createTable(this.tableName, (table) => {
      table.increments('id')
      table.enum('status', Object.values(PostStatus))
        .defaultTo(PostStatus.DRAFT)
        .notNullable()
      // ...
    })
  }

  public async down () {
    this.schema.dropTable(this.tableName)
  }
}
```

Then, I import the enum into model file and use it to typing the column:

```ts
// ./app/Models/Post.ts 
import { BaseModel, column } from '@ioc:Adonis/Lucid/Orm'
import { PostStatus } from 'Contracts/enums'

export default class Post extends BaseModel {
  @column({ isPrimary: true })
  public id: number

  @column()
  public status: PostStatus

  // ...
}
```

Now, you only need to update the list in one place. It's more organized, for me.

The only "downside" is that u will have to use this enum each time u create a post, update "status" field or write a query.

```ts
// ./app/Controllers/Http/PostsController.ts
import Post from 'App/Models/Post'
import { PostStatus } from 'Contracts/enums'

export default class PostsController {
  public async index() {
    const posts = await Post.query()
      .where('status', PostStatus.PUBLIC)
  
    return posts
  }

  public async create() {
    const post = await Post.create({
      title: 'lorem ipsum dolor...',
      status: PostStatus.PRIVATE,
    })

    return post
  }

  public async update({ params }) {
    const post = await Post.findOrFail(params.id)

    post.status = PostStatus.PUBLIC
    await post.save()
  
    return post
  }
}
```

**Comment by RomainLanz** [comment]
Hey! 👋 

You can use String Union or a TypeScript Enum.

**Via String Union:**
```ts
@column()
public type: 'Rodada' | 'Mensal' // ...
```

**Via Enum:**
```ts
enum LeagueType {
  Rodada = 'Rodada',
  Mensal = 'Mensal',
  // ...
}

//

@column()
public type: LeagueType
```

**Reply by arthurcvm** [reply]
Hi @RomainLanz, I use String type in model and enum in validator, works for me.

In model:
```typescript
  @column()
  public type: string
```
In validator:
```typescript
type: schema.enum(['page', 'route'])
```

Return of validator with wrong data:
```json
{
  "errors": [
    {
      "rule": "enum",
      "field": "type",
      "message": "enum validation failed",
      "args": {
        "choices": [
          "page",
          "route"
        ]
      }
    }
  ]
}
```

Following your sugestion, via String Union, crash my validator.

**Reply by thetutlage** [reply]
Validator has nothing to do with the model. They are two different modules all together. Also, you can read about `enum` types in Typescript

**Reply by arthurcvm** [reply]
Yes, but my validator is separate from the model, I sent these excerpts just as an example of how it is working in my app, where the validation is working correctly.

**Reply by ialiaslani** [reply]
What I Did Was That I Created An Enam 
```
export enum Priority {
  NORMAL = 'normal',
  HIGH = 'high',
  URGENT = 'urgent',
}

```
and Validated Like 
`priority: schema.enum(["normal", "high", "urgent"])`

And Then Saved Data Like 
```

 priority: Priority.URGENT !== validatedData.priority? Priority.HIGH !== validatedData.priority ?
            Priority.NORMAL : Priority.HIGH : Priority.URGENT
```

You Can Use Switch Case Too In An Extra Function
